### ISSUE TITLE
using .set on nested field in pre-save hook overwrites parent when using .select

### ISSUE THREAD

**Issue by zoellner** [Automattic/mongoose#5800]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
Parent field gets overwritten by .set of nested field in pre-save hook when document is incomplete due to .select

**If the current behavior is a bug, please provide the steps to reproduce.**
https://gist.github.com/zoellner/f022ac62a7ed968a85784edf1a6690c8

**What is the expected behavior?**
Nested field gets updated without overwriting the parent field.

**Please mention your node.js, mongoose and MongoDB version.**
node.js 6.11.1
mongoose 4.13.0
mongoDB 3.4.7

**Comment by vkarpov15** [comment]
Thanks for the detailed repro, will fix asap :+1:

**Comment by dietergeerts** [comment]
I just encounter this bug in version 5.9.16. Does this mean that this got reverted, and when doing changes in a sub-document, it has to be selected in the first place?

I wrote a plugin that does some extra changes based on other changes. The plugin is sure that the sub-documents exist, and it wants to update 1 specific field. Though using `this.set` completely overwrites the whole sub-document, which is not expected. Even though it was not specifically selected, and it conforms to the schema, I would expect only that path to be updated, not the whole sub-document. I must say that this is in a sub-document array.
